### PR TITLE
ALTAPPS-1155: Update subscription state after expiration

### PR DESCRIPTION
**Pull request by XanderZhu**
**YouTrack Issues**:
[ALTAPPS-1155](https://vyahhi.myjetbrains.com/youtrack/issue/ALTAPPS-1155)

**Checklist**

_Before Code Review:_

- [x] Fields "Assignees, Labels, Milestone" are filled in the pull request;
- [x] All checks have been passed;
- [x] Changes have been checked locally;
- [x] Saved in the cache models serialization checked locally.

**Description**
- Apply freemium limits to the expired mobile-only subscription;
- Fetch subscription from remote on app startup, if the cached subscription is expired mobile-only;
- Wait for mobile-only subscription expiration and then sync the subscription.

### CODE REVIEW

**Comment by XanderZhu** [comment]
@ivan-magda 
The iOS build failed, but it was successfully built locally.

File: shared/src/commonMain/kotlin/org/hyperskill/app/main/presentation/AppActionDispatcher.kt [hunk 1]
```diff
@@ -67,10 +72,11 @@ internal class AppActionDispatcher(
         if (isSubscriptionPurchaseEnabled) {
             currentSubscriptionStateRepository
                 .changes
-                .map { it.type }
-                .distinctUntilChanged()
-                .onEach { subscriptionType ->
-                    onNewMessage(AppFeature.InternalMessage.SubscriptionTypeChanged(subscriptionType))
+                .distinctUntilChangedBy {
+                    Triple(it.type, it.status, it.validTill)
```
Comment (ivan-magda) [inline shared/src/commonMain/kotlin/org/hyperskill/app/main/presentation/AppActionDispatcher.kt:76]: For me it's unclear why we should observe only changes in `type`, `status` and `validTill` of the subscription, why not use `distinctUntilChanged()` on whole subscription model itself?

File: shared/src/commonMain/kotlin/org/hyperskill/app/manage_subscription/presentation/ManageSubscriptionReducer.kt [hunk 2]
```diff
@@ -50,12 +55,36 @@ internal class ManageSubscriptionReducer : StateReducer<State, Message, Action>
...
-    private fun handleManageSubscriptionClicked(state: State): ReducerResult =
-        if (state is State.Content && state.manageSubscriptionUrl != null) {
-            state to setOf(
-                InternalAction.LogAnalyticsEvent(ManageSubscriptionClickedManageHyperskillAnalyticEvent),
-                Action.ViewAction.OpenUrl(state.manageSubscriptionUrl)
-            )
+    private fun handleActionButtonClicked(state: State): ReducerResult =
+        if (state is State.Content) {
+            state to when {
+                state.subscription.isActive && state.manageSubscriptionUrl != null -> {
```
Comment (ivan-magda) [inline shared/src/commonMain/kotlin/org/hyperskill/app/manage_subscription/presentation/ManageSubscriptionReducer.kt:61]: `ManageSubscriptionViewStateMapper` has the same condition, you can extract it to state extension

File: shared/src/commonMain/kotlin/org/hyperskill/app/subscriptions/domain/interactor/SubscriptionsInteractor.kt [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+                currentProfileStateRepository.updateState {
+                    it.copy(gamification = it.gamification.copy(passedProblems = it.gamification.passedProblems + 1))
+                }
+            }
+        }
+    }
+
+    suspend fun refreshSubscriptionOnExpirationIfNeeded(subscription: Subscription) {
+        refreshMobileOnlySubscriptionJob?.cancel()
+        refreshMobileOnlySubscriptionJob = null
```
Comment (ivan-magda) [inline shared/src/commonMain/kotlin/org/hyperskill/app/subscriptions/domain/interactor/SubscriptionsInteractor.kt:67]: `cancelSubscriptionRefresh()`

**Review by ivan-magda — APPROVED** [review]

File: shared/src/commonMain/kotlin/org/hyperskill/app/main/presentation/AppActionDispatcher.kt (hunk 1 above)
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/main/presentation/AppActionDispatcher.kt:76]: I used filtration by `type`, `status`, and `validTill` fields to reduce sending the `SubscriptionChanged` message.
I agree, that it is redundant in this case.

**Comment by XanderZhu** [comment]
@ivan-magda 
Please, take a look at the latest changes.
https://vyahhi.myjetbrains.com/youtrack/issue/ALTAPPS-1155/Shared-Android-User-has-to-restart-application-after-mobile-subscription-cancellation#focus=Comments-74-282878.0-0

File: shared/src/commonMain/kotlin/org/hyperskill/app/subscriptions/domain/interactor/SubscriptionsInteractor.kt [hunk 4]
```diff
@@ -86,18 +86,19 @@ class SubscriptionsInteractor(
         val nowByUTC = Clock.System.now()
             .toLocalDateTime(TimeZone.UTC)
             .toInstant(TimeZone.UTC)
-        val delayDuration = subscriptionValidTill - nowByUTC
+
+        // Add one minute to wait until the subscription is synced on the backend
```
Comment (ivan-magda) [inline shared/src/commonMain/kotlin/org/hyperskill/app/subscriptions/domain/interactor/SubscriptionsInteractor.kt:90]:
```suggestion
        // ALTAPPS-1155: Add one minute to wait until the subscription is synced on the backend
```

**Pull request merged**